core/platform/lf_rp2040_support.c

```diff
@@ -0,0 +1,181 @@
+    sem_init(&_lf_sem_irq_event, 0, 1);
+}
+
+/**
+ * Fetch the value of an internal (and platform-specific) physical clock and
+ * store it in `t`. in nanoseconds
+ *
+ * Ideally, the underlying platform clock should be monotonic. However, the
+ * core lib tries to enforce monotonicity at higher level APIs (see tag.h).
+ * TODO: might want to use the RTC
+ * @return 0 for success, or -1 for failure
+ */
+int _lf_clock_now(instant_t* t) {
+    absolute_time_t now;
+    uint64_t ns_from_boot;
+    now = get_absolute_time();
+    ns_from_boot = to_us_since_boot(now) * 1000;
+    *t = (instant_t) ns_from_boot;
+    return 0; 
+}
+
+/**
+ * Pause execution for a given duration.
+ * 
+ * @return 0 for success, or -1 for failure.
+ */
+int lf_sleep(interval_t sleep_duration) {
+    if (sleep_duration < 0) {
+        return -1;
+    }
+    sleep_us((uint64_t) (sleep_duration / 1000));
+    return 0;
+}
+
+/**
+ * @brief Sleep until the given wakeup time.
+ * 
+ * @param wakeup_time The time instant at which to wake up.
+ * @return int 0 if sleep completed, or -1 if it was interrupted.
+ */
+int _lf_interruptable_sleep_until_locked(environment_t* env, instant_t wakeup_time) {
+    int ret_code = 0;
+    if (wakeup_time < 0) {
+        ret_code = -1;
+        return ret_code;
+    }
+    absolute_time_t target;
+    // reset semaphore to 0
+    // TODO: leverage the semaphore permit number 
+    sem_reset(&_lf_sem_irq_event, 0);
+    target = from_us_since_boot((uint64_t) (wakeup_time / 1000));
+    lf_critical_section_exit(env);
+    // sleep till target or return on processor event
+    if(sem_acquire_block_until(&_lf_sem_irq_event, target)) {
+        ret_code = -1;
+    }
+    lf_critical_section_enter(env);
+    return ret_code;
+}
+/*
+* Critical sections are only provided for an unthreaded, single core
+* runtime. In the unthreaded runtime, all interactions with core1 are through
+* physical actions and interupts outside of the runtime.
+*/
+#ifdef LF_UNTHREADED
+
+/**
+ * Enter a critical section where logical time and the event queue are guaranteed
+ * to not change unless they are changed within the critical section.
+ * this can be implemented by disabling interrupts.
+ * Users of this function must ensure that lf_init_critical_sections() is
+ * called first and that lf_critical_section_exit() is called later.
+ * @return 0 on success, platform-specific error number otherwise.
+ * TODO: needs to be used sparingly 
+ */
+int lf_disable_interrupts_nested() {
+    if (!critical_section_is_initialized(&_lf_crit_sec)) {
+        return 1;
+    } 
+    // disables irq and spin-locks core
+    if (_lf_num_nested_critical_sections++ == 0) {
+        critical_section_enter_blocking(&_lf_crit_sec);
+    }
+    return 0;
+}
+
+/**
+ * Exit the critical section entered with lf_lock_time().
+ * @return 0 on success, platform-specific error number otherwise.
+ * TODO: needs to be used sparingly, find a better way for event queue
```

Comment:
All of these TODOs are basically the same as FIXMEs in that they probably should be moved out of the code and into GitHub issues before this is merged.